Sources/Views/EpisodePage.swift

```diff
@@ -0,0 +1,673 @@
+  guard user == nil else { return [] }
+
+  return [
+    .span(attributes: [.class([Class.padding([.mobile: [.left: 2]])])], "or"),
+    .a(
+      attributes: [
+        .href(path(to: .login(redirect: url(to: .episode(.left(ep.slug)))))),
+        .class([Class.pf.components.button(color: .black, style: .underline)])
+      ],
+      "Log in"
+    )
+  ]
+}
+
+private func episodeInfoView(
+  permission: EpisodePermission,
+  ep: Episode,
+  previousEpisodes: [Episode],
+  date: () -> Date
+  ) -> Node {
+  return .div(
+    attributes: [.class([Class.padding([.mobile: [.all: 3], .desktop: [.all: 4]]), Class.pf.colors.bg.white])],
+    topLevelEpisodeInfoView(episode: ep, date: date),
+    previousEpisodesView(of: ep, previousEpisodes: previousEpisodes),
+    sectionsMenu(episode: ep, permission: permission)
+  )
+}
+
+private func previousEpisodesView(of ep: Episode, previousEpisodes: [Episode]) -> Node {
+  guard !previousEpisodes.isEmpty else { return [] }
+
+  return [
+    .p(
+      attributes: [.class([Class.padding([.mobile: [.top: 1], .desktop: [.top: 1]]), Class.pf.colors.bg.white, Class.pf.type.body.leading])],
+      "This episode builds on concepts introduced previously:"
+    ),
+    .ul(
+      attributes: [.class([
+        Class.type.list.styleNone,
+        Class.padding([.mobile: [.left: 2]])
+        ])],
+      .fragment(
+        previousEpisodes.map {
+          .li(
+            attributes: [.class([Class.pf.type.body.leading])],
+            "#",
+            .text(String($0.sequence)),
+            ": ",
+            .a(
+              attributes: [
+                .class([Class.pf.colors.link.purple]),
+                .href(url(to: .episode(.left($0.slug))))
+              ],
+              .text($0.title)
+            )
+          )
+        }
+      )
+    )
+  ]
+}
+
+private func sectionsMenu(episode: Episode, permission: EpisodePermission?) -> Node {
+  guard let permission = permission, isEpisodeViewable(for: permission) else { return [] }
+
+  let exercisesNode: Node? = episode.exercises.isEmpty
+    ? nil
+    : .a(attributes: [.class([Class.pf.colors.link.purple, Class.margin([.mobile: [.right: 2]])]), .href("#exercises")],
+        "Exercises")
+
+  let referencesNode: Node? = episode.references.isEmpty
+    ? nil
+    : .a(attributes: [.class([Class.pf.colors.link.purple, Class.margin([.mobile: [.right: 2]])]), .href("#references")],
+        "References")
+
+  // Don't show quick link menu if at least one of exercises or references are present.
+  guard exercisesNode != nil || referencesNode != nil else { return [] }
+
+  return .div(
+    attributes: [.class([Class.padding([.mobile: [.top: 2], .desktop: [.top: 3]])])],
+    .fragment([
+      .a(
+        attributes: [.class([Class.pf.colors.link.purple, Class.margin([.mobile: [.right: 2]])]), .href("#transcript")],
+        "Transcript"
+      ),
+      exercisesNode,
+      referencesNode
+      ]
+      .compactMap(id)
```

Comment:
Could we get rid of this `compactMap` and `fragment` by using `Node` above instead of `Node?`?